### PR TITLE
[FTE] Adaptive reorder partitioned join

### PR DESCRIPTION
**Pull request by gaurav8297**
## Description
Flip sides for partitioned join based on data size stats. This rule is only used during Adaptive Planning in FTE.

It helps fix the join order due to either missing or incorrect stats, or maybe wrong estimations caused by skewed distribution of filters, etc.

### Benchmarks
Example query using the TPCDS dataset. Here the last join has an incorrect order.
```
SELECT sum(c.c_current_cdemo_sk), max(c.c_birth_country),
sum(ss.ss_quantity), sum(ss.ss_list_price), sum(ss.ss_coupon_amt),
max(i.i_current_price), sum(cs.cs_wholesale_cost), sum(cs.cs_list_price),
sum(cs.cs_sales_price), sum(cs.cs_ext_sales_price),
sum(cs.cs_net_paid_inc_tax), sum(cs.cs_net_paid_inc_ship),
sum(cs.cs_net_profit), sum(cs.cs_ext_tax), sum(cs.cs_coupon_amt),
sum(cs.cs_ext_ship_cost), sum(cs.cs_net_paid), sum(cs.cs_net_paid_inc_tax),
sum(cs.cs_net_paid_inc_ship), sum(cs.cs_call_center_sk),
sum(cs.cs_net_paid_inc_ship_tax), sum(cs.cs_net_profit),
sum(cs.cs_call_center_sk), sum(cs.cs_warehouse_sk), sum(cs.cs_bill_hdemo_sk)
FROM store_sales AS ss
LEFT JOIN item AS i
ON ss.ss_item_sk = i.i_item_sk
LEFT JOIN customer AS c ON (ss.ss_customer_sk = c.c_customer_sk)
LEFT JOIN catalog_sales AS cs  ON (ss.ss_item_sk=cs.cs_item_sk AND ss.ss_sold_date_sk = cs.cs_sold_date_sk)
WHERE c.c_birth_year > 1968;
```
Before:
```
Query 20240805_175839_00010_wvcdc, FINISHED, 8 nodes
Splits: 13,798 total, 13,798 done (100.00%)
2:05 [4.33B rows, 68.7GB] [34.7M rows/s, 564MB/s]
```
After:
```
Query 20240805_175654_00008_wvcdc, FINISHED, 8 nodes
Splits: 9,394 total, 9,394 done (100.00%)
1:12 [4.33B rows, 68.7GB] [60.5M rows/s, 982MB/s]
```

There’s also a significant improvement in CPU and memory usage.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# General
* Add optimisation which can reorder joins at runtime if retry_policy is set to `TASK`, even if table statistics are not available.
  Note: In case of unexpected issues feature can be disabled with  `fault_tolerant_execution_adaptive_join_reordering_enabled` session property
  or fault-tolerant-execution-adaptive-join-reordering-enabled config property. ({issue}`23182`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/AdaptiveReorderPartitionedJoin.java [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    // Use a smaller threshold to change the order since the cost of changing the order is lower here. Additionally,
+    // the data size stats are more accurate compared to static planning since they are collected at run time from the
+    // stage execution.
+    private static final double SIZE_DIFFERENCE_THRESHOLD = 1.5;
+
+    // With speculative execution, reordering small joins might cause unnecessary restarts of the
+    // join stage and lead to performance degradation. Hence, we only reorder if the size of the right side is
+    // above this threshold.
+    // TODO: Consider the cost of restarting the stage as part of adaptive planning.
+    private static final double MIN_SIZE_THRESHOLD = DataSize.of(5, GIGABYTE).toBytes();
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/AdaptiveReorderPartitionedJoin.java:109]: can we make those configurable (or at least threshold) so we can do some experiments what works best?

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestAdaptiveReorderPartitionedJoin.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                                        ImmutableSet.of("probeSymbol"),
+                                        Optional.of(ImmutableList.of(ImmutableList.of("probeSymbol", "symbol2"))),
+                                        remoteSource(
+                                                ImmutableList.of(new PlanFragmentId("1")),
+                                                ImmutableList.of("probeSymbol", "symbol2"),
+                                                REPARTITION)))));
+    }
+
+    @Test
+    public void testNoChangesWhenBuildSourceIsSmaller()
```
Comment (losipiuk) [inline core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestAdaptiveReorderPartitionedJoin.java:183]: you can also test the threshold, and size_limit logic.

**Review by losipiuk — COMMENTED** [review]
Thanks - generally looks fine. Minor suggestions.
As this is disabled by default we can merge it soon and tweak it out.

**Review by losipiuk — APPROVED** [review]

**Comment by mosabua** [comment]
@gaurav8297 @losipiuk are you sending docs? Minimal content is probably what the release notes say ...

**Comment by losipiuk** [comment]
> @gaurav8297 @losipiuk are you sending docs? Minimal content is probably what the release notes say ...

Yeah - we should add something. @gaurav8297 can you send a small PR some time before next Trino release?

**Comment by mosabua** [comment]
@gaurav8297 .. we are aiming to release tomorrow .. can you send a PR soon?

**Pull request merged**